ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/LedgerInitialization.scala

```diff
@@ -0,0 +1,159 @@
+// Copyright (c) 2020 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.ledger.participant.state.kvutils
+
+import java.time.{Duration, Instant}
+import java.util.UUID
+
+import akka.NotUsed
+import akka.stream.Materializer
+import akka.stream.scaladsl.{Sink, Source}
+import com.daml.dec.{DirectExecutionContext => DEC}
+import com.daml.ledger.participant.state.v1.Update.{ConfigurationChangeRejected, ConfigurationChanged}
+import com.daml.ledger.participant.state.v1.{Configuration, Offset, ReadService, SubmissionId, SubmissionResult, Update, WriteService}
+import com.daml.lf.data.Time.Timestamp
+import org.slf4j.LoggerFactory
+
+import scala.compat.java8.FutureConverters._
+import scala.concurrent.Future
+import scala.concurrent.duration.DurationInt
+import scala.util.{Failure, Success}
+
+/**
+ *
+ * @param maxUpdates Maximum number of ReadService updates to inspect
+ * @param initialDelay Delay until the first config change is submitted
+ * @param retryDelay Delay between submission retries
+ * @param submissionTimeout Timeout until a submission is declared lost
+ * @param maxAttempts Maximum number of submission attempts
+ */
+case class LedgerInitializationConfig(
+  maxUpdates: Long,
+  initialDelay: Duration,
+  retryDelay: Duration,
+  submissionTimeout: Duration,
+  maxAttempts: Long,
+)
+
+object LedgerInitializationConfig {
+  def reasonableDefault: LedgerInitializationConfig = LedgerInitializationConfig(
+    maxUpdates = 10,
+    initialDelay = Duration.ofSeconds(2),
+    retryDelay = Duration.ofSeconds(1),
+    submissionTimeout = Duration.ofSeconds(60),
+    maxAttempts = 5,
+  )
+}
+
+object LedgerInitialization {
+
+  private val logger = LoggerFactory.getLogger(this.getClass)
+
+  def initialize(
+      readService: ReadService,
+      writeService: WriteService,
+      initialConfig: Configuration,
+      config: LedgerInitializationConfig = LedgerInitializationConfig.reasonableDefault,
+  )(implicit mat: Materializer): Future[Configuration] = {
+
+    // Mutable state
+    object State {
+      var submittedId: Option[SubmissionId] = None
+      var attempt = 1
+      var skippedUpdates = 0
+      var retryDelay: Duration = config.retryDelay
+      var nextSubmission: Instant = Instant.now.plus(config.initialDelay)
+      var submissionTimeout: Instant = Instant.EPOCH
+
+      def isCurrentSubmission(submissionId: SubmissionId): Boolean =
+        submittedId.contains(submissionId)
+
+      def shouldSubmitNow(): Boolean = {
+        val now = Instant.now
+        (submittedId.isEmpty && now.isAfter(nextSubmission)) || (submittedId.isDefined && now
+          .isAfter(submissionTimeout))
+      }
+
+      def scheduleRetry(reason: String): Future[Option[Configuration]] = {
```

Comment:
Maybe an enumeration rather than a `String`? Means the logging is handled in one place, not several.